include/xtensor/xbuilder.hpp

```diff
@@ -69,6 +70,167 @@ namespace xt
         return zeros<T>(std::vector<I>(shape));
     }
 
+    namespace detail {
+        template <class Functor>
+        inline auto make_xindex_function(Functor&& f, std::initializer_list<std::size_t> shape) noexcept
+        {
+            using type = xindex_function<Functor, typename Functor::value_type>;
+            return type(f, std::forward<std::vector<std::size_t>>(std::vector<std::size_t>(shape)));
+        }
+        
+        template <class Functor, class S>
+        inline auto make_xindex_function(Functor&& f, const S& shape) noexcept
+        {
+            using type = xindex_function<Functor, typename Functor::value_type>;
+            return type(f, std::forward<S>(shape));
+        }
+
+        template <class T>
+        struct arange_impl {
+            T m_start, m_stop, m_step;
+            using value_type = T;
+            arange_impl(T start, T stop, T step) : 
+                m_start(start), m_stop(stop), m_step(step) {
+            }
+
+            template <class... Args>
+            T operator()(Args... args) const {
+                std::vector<T> args_arr({static_cast<T>(args)...});
+                return m_start + m_step * args_arr[0];
+            }
+
+            T operator[](const xindex& idx) const {
+                return m_start + m_step * idx[0];
+            }
+        };
+
+        template <class T>
+        struct repeat_impl {
+            using value_type = typename T::value_type;
+            using size_type = typename T::size_type;
+            using return_type = T;
+            
+            const T& m_V;
+            const size_type m_axis;
+
+            repeat_impl(const T& V, const size_type axis) :
+                m_V(V), m_axis(axis) {
+            }
+
+            template <class... Args>
+            value_type operator()(Args... args) const {
+                std::vector<size_type> args_arr({static_cast<size_type>(args)...});
+                return m_V(args_arr[m_axis]);
```

Comment:
Actually no, since apply will allocate an array on the stack (if I remember well); your proposal is better because much simpler :)